pkg/storaged/crypto-keyslots.jsx

```diff
@@ -0,0 +1,556 @@
+                          else
+                              return passphrase_remove(block, vals.passphrase);
+                      }
+                  }
+    });
+}
+
+const RemoveClevisField = (tag, key, dev) => {
+    return {
+        tag: tag,
+        title: <span className="pficon pficon-warning-triangle-o large" />,
+        options: { },
+        initial_value: "",
+
+        render: (val, change) => {
+            return (
+                <div data-field={tag}>
+                    <div className="large">{ fmt_to_array(_("Remove $0?"), <b>{key.url}</b>) }</div>
+                    <div>{ fmt_to_array(_("Keyserver removal may prevent unlocking $0."), <b>{dev}</b>) }</div>
+                </div>
+            );
+        }
+    };
+};
+
+function remove_clevis_dialog(client, block, key) {
+    dialog_open({ Title: _("Remove Tang keyserver"),
+                  Fields: [
+                      RemoveClevisField("keyserver", key, block_name(block))
+                  ],
+                  Action: {
+                      DangerButton: true,
+                      Title: _("Remove"),
+                      action: function () {
+                          return clevis_remove(block, key);
+                      }
+                  }
+    });
+}
+
+export class CryptoKeyslots extends React.Component {
+    constructor() {
+        super();
+        this.state = { version: 1, slots: null, slot_error: null, max_slots: 8 };
+    }
+
+    monitor_slots(block) {
+        // HACK - we only need this until UDisks2 has a Encrypted.Slots property or similar.
+        if (block != this.monitored_block) {
+            if (this.monitored_block)
+                this.monitor_channel.close();
+            this.monitored_block = block;
+            if (block) {
+                var dev = decode_filename(block.Device);
+                this.monitor_channel = python.spawn(luksmeta_monitor_hack_py, [ dev ],
+                                                    { superuser: true });
+                var buf = "";
+                this.monitor_channel.stream(output => {
+                    var lines;
+                    buf += output;
+                    lines = buf.split("\n");
+                    buf = lines[lines.length - 1];
+                    if (lines.length >= 2) {
+                        const data = JSON.parse(lines[lines.length - 2]);
+                        this.setState({ slots: data.slots, version: data.version });
+                    }
+                });
+                this.monitor_channel.fail(err => {
+                    this.setState({ slots: [ ], slot_error: err });
+                });
+            }
+        }
+    }
+
+    componentDidUmount() {
+        this.monitor_slots(null);
+    }
+
+    render() {
+        var client = this.props.client;
+        var block = this.props.block;
+
+        if (!client.features.clevis)
+            return null;
+
+        this.monitor_slots(block);
```

Comment:
`monitor_slots` should be pretty cheap if it's already monitored. But especially if you need to make the state more complex (like the `block → channel` map), it might make more sense to move this to `componentWillMount()`, instead of calling this all the time in `render()`.

Reply:
I think `componentWillMount` does not see all property changes, no?

Reply:
You mean if a new block objects appear while the component is mounted? True that, so ignore this. But if that is/can be the case, then the logic of always just monitoring one block object instead of a map is even more confusing.

Reply:
> You mean if a new block objects appear while the component is mounted?

For example.  I am not that this can _actually_ happen given the way that the CryptoKeyslots instances are used, but we should support it nonetheless, to reduce surprises.